File: client/src/main/java/net/treset/minecraftlauncher/news/NewsService.java
package net.treset.minecraftlauncher.news;

import javafx.util.Pair;
import net.treset.mc_version_loader.json.SerializationException;
import net.treset.minecraftlauncher.LauncherApplication;
import net.treset.minecraftlauncher.util.HttpService;

import java.io.IOException;

public class NewsService extends HttpService {
    public NewsService() {
        super(LauncherApplication.config.UPDATE_URL);
    }

    public News news() throws IOException {
        Pair<HttpStatusCode, byte[]> result = get("news", LauncherApplication.stringLocalizer.get("launcher.version"), LauncherApplication.stringLocalizer.getLanguage().getLocale());
        String response = new String(result.getValue());
        try {
            return News.fromJson(response);
        } catch(SerializationException e) {
            throw new IOException("Failed to parse the response from the server.\nError: " + e);
        }
    }
}

File: client/composeApp/src/commonMain/kotlin/net/treset/treelauncher/backend/news/NewsService.kt
package net.treset.treelauncher.backend.news

import net.treset.mcdl.json.SerializationException
import net.treset.treelauncher.backend.util.HttpService
import net.treset.treelauncher.localization.language
import net.treset.treelauncher.localization.strings
import java.io.IOException

class NewsService(url: String) : HttpService(url) {
    @Throws(IOException::class)
    fun news(): News {
        val result: Pair<HttpStatusCode, ByteArray> = get(
            "news",
            strings().launcher.version(),
            language().appLanguage.locale
        )
        val response = String(result.second)
        return try {
            News.fromJson(response)
        } catch (e: SerializationException) {
            throw IOException("Failed to parse the response from the server.\nError: $e")
        }
    }
}
